dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/system/SystemUpdateService.java

```diff
@@ -0,0 +1,302 @@
+                    }
+                }
+            }
+        }
+
+        return newerPatchVersions;
+    }
+
+    /**
+     * Converts a list of json elements representing patch versions, into a map
+     * of: SemVer (parsed version) and a map of strings representing the message
+     * to send to the admin user(s).
+     */
+    protected static Map<Semver, Map<String, String>> convertJsonToMap( List<JsonElement> patchVersions )
+    {
+        Map<Semver, Map<String, String>> versionsAndMessage = new TreeMap<>();
+
+        for ( JsonElement patchVersion : patchVersions )
+        {
+            JsonObject patchJsonObject = patchVersion.getAsJsonObject();
+
+            String version = patchJsonObject.getAsJsonPrimitive( FIELD_NAME_NAME ).getAsString();
+
+            Map<String, String> message = new HashMap<>();
+            message.put( FIELD_NAME_VERSION, patchJsonObject.getAsJsonPrimitive( FIELD_NAME_NAME ).getAsString() );
+            message.put( FIELD_NAME_RELEASE_DATE,
+                patchJsonObject.getAsJsonPrimitive( FIELD_NAME_RELEASE_DATE ).getAsString() );
+            message.put( FIELD_NAME_DOWNLOAD_URL, patchJsonObject.getAsJsonPrimitive( FIELD_NAME_URL ).getAsString() );
+
+            versionsAndMessage.put( new Semver( version ), message );
+        }
+
+        return versionsAndMessage;
+    }
+
+    public void sendMessageForEachVersion( Map<Semver, Map<String, String>> patchVersions )
+    {
+        Set<User> recipients = getRecipients();
+
+        for ( Map.Entry<Semver, Map<String, String>> entry : patchVersions.entrySet() )
+        {
+            Semver version = entry.getKey();
+            Map<String, String> message = entry.getValue();
+
+            for ( User recipient : recipients )
+            {
+                // Check if message has been sent before using
+                // version.getValue() as extMessageId
+                List<MessageConversation> existingMessages = messageService.getMatchingExtId( version.getValue() );
+
+                if ( existingMessages.isEmpty() )
+                {
+                    MessageConversationParams params = new MessageConversationParams.Builder()
+                        .withRecipients( ImmutableSet.of( recipient ) )
+                        .withSubject( NEW_VERSION_AVAILABLE_MESSAGE_SUBJECT )
+                        .withText( buildMessageText( message ) )
+                        .withMessageType( MessageType.SYSTEM )
+                        .withExtMessageId( version.getValue() ).build();
+
+                    messageService.sendMessage( params );
+                }
+            }
+        }
+    }
+
+    private Set<User> getRecipients()
+    {
+        Set<User> recipients = messageService.getSystemUpdateNotificationRecipients();
+
+        // Fallback to fetching all users with ALL authority for our recipient
+        // list.
+        if ( recipients.isEmpty() )
+        {
+            recipients = getUsersWithAllAuthority();
```

Comment:
I like to avoid such "in place" variables that need mental tracking even if it is just to the next statement.
Maybe use a `return` instead or do a `return !recipients.isEmpty() ? recipients : getUsersWithAllAuthority();`?